pkg/tnf/operator/starter.go

```diff
@@ -46,11 +48,60 @@ var (
 	fencingUpdateMutex sync.Mutex
 )
 
+type DualReplicaClusterHandler struct {
+	ctx           context.Context
+	kubeClient    kubernetes.Interface
+	clusterStatus status.ExternalEtcdClusterStatus
+}
+
+func NewDualReplicaClusterHandler(ctx context.Context,
+	opClient v1helpers.StaticPodOperatorClient,
+	kubeClient kubernetes.Interface,
+	configInformers configv1informers.SharedInformerFactory) (*DualReplicaClusterHandler, error) {
+
+	bootstrapCompleted := false
+	readyForEtcdRemoval := false
+	dualReplicaEnabled, err := ceohelpers.IsDualReplicaTopology(ctx, configInformers.Config().V1().Infrastructures().Lister())
+	if err != nil {
+		klog.Errorf("failed to determine DualReplicaTopology, aborting controller start: %v", err)
+		return nil, err
+	} else if dualReplicaEnabled {
+		klog.Infof("detected DualReplica topology")
+
+		// Check if bootstrap is completed without waiting for it.
+		// If it's completed already, CEO probably was restarted for an update.
+		_, opStatus, _, err := opClient.GetStaticPodOperatorState()
+		if err != nil {
+			klog.Errorf("failed to get static pod operator state: %v", err)
+			return nil, err
+		} else {
```

Comment:
This else statement is also implied, since you return if you can't find the operator state.